// letsbonkdotfun-sdk/src/constants/programs.ts
import { PublicKey } from '@solana/web3.js';

/**
 * Program IDs - Core protocol addresses
 */
export const LETSBONK_PROGRAM_ID = new PublicKey('LanMV9sAd7wArD4vJFi2qDdfnVhFxYSUg6eADduJ3uj');
export const RAYDIUM_AUTHORITY = new PublicKey('WLHv2UAZm6z4KyaaELi5pjdbJh6RESMva1Rnn8pJVVh');
export const GLOBAL_CONFIG = new PublicKey('6s1xP3hpbAfFoNtUNF8mfHsjr2Bd97JxFJRWLbL6aHuX');
export const PLATFORM_CONFIG = new PublicKey('FfYek5vEz23cMkWsdJwG2oa6EphsvXSHrGpdALN4g6W1');
export const EVENT_AUTHORITY = new PublicKey('2DPAtwB8L12vrMRExbLuyGnC7n2J5LNoZQSejeQGpwkr');

/**
 * Standard Solana program addresses
 */
export const TOKEN_PROGRAM_ID = new PublicKey('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
export const ASSOCIATED_TOKEN_PROGRAM_ID = new PublicKey(
  'ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL'
);
export const SYSTEM_PROGRAM_ID = new PublicKey('11111111111111111111111111111111');
export const METAPLEX_PROGRAM = new PublicKey('metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s');
export const RENT_SYSVAR = new PublicKey('SysvarRent111111111111111111111111111111111');

/**
 * Token addresses
 */
export const WSOL_TOKEN = new PublicKey('So11111111111111111111111111111111111111112');

/**
 * Platform admin address
 */
export const PLATFORM_ADMIN = new PublicKey('2P56vRWDrCBGkqYXxgSWAnuZQZrJPySRQGToTJThpmkN');
